Optimize torch zeros

edvgha opened the pull request:
Summary: After creating empty tensor 'memset' used to zero out items of tensor

Test Plan:
1. Use scripts/arvindkannan/benchmark.py
2. TestZeros in basic.cpp

OMP_NUM_THREADS=1 MKL_NUM_THREADS=1  buck test mode/opt caffe2/aten:basic -- --print-passing-details BasicTest.BasicTestCPU

TestZeros will print runtime for torch.zeros in milliseconds .

Before: 50975
After:    43469

pytorch benchmark tool results:

timer = benchmark_utils.Timer(stmt="torch.zeros((1024, 4096))")

Before: 1007 us
After:     841.26 us
1 measurement, 10000 runs , 1 thread

Differential Revision: D23925113

robieta commented:
I put together a more comprehensive benchmark:

(This assumes two conda envs: `45636_ref` built from `f6dc256bc65e80f48f07f497b12ea01ed8d78e34` and `45636_pr` built from this PR)
benchmark_45636.py
```
import os
import pickle
import sys

import torch
from torch.utils.benchmark import Compare, Timer
from torch.utils.benchmark.op_fuzzers import unary

MIN_RUN_TIME = 1
METADATA = {
    "45636_ref": (
        "Reference",
        "/tmp/45636_ref.pkl",
    ),
    "45636_pr": (
        "PR #45636",
        "/tmp/45636_pr.pkl"
    ),
}


def measure_main():
    env = os.getenv("CONDA_PREFIX")
    env = os.path.split(env)[1]
    assert env in METADATA

    results = []
    print("Starting time measurements.")
    for dtype in [torch.int8, torch.int32, torch.float16, torch.float32]:
        for num_threads in [1, 2, 4]:
            for n in [1, 64, 1024, 128 * 1024, 123456789]:
                timer = Timer(
                    "x.zero_()",
                    globals={"x": torch.ones((n,), dtype=dtype)},
                    num_threads=num_threads,
                    env=METADATA[env][0],
                    label=f"x.zero_()  {dtype}",
                    description=str(n),
                )
                results.append(timer.blocked_autorange(min_run_time=MIN_RUN_TIME))
                print(".", end="")
                sys.stdout.flush()

            for tensors, _, _ in unary.UnaryOpFuzzer(seed=0, dtype=dtype).take(10):
                if not tensors["x"].is_contiguous():
                    continue  # This PR is a contiguous fast path.

                timer = Timer(
                    "x.zero_()",
                    globals=tensors,
                    num_threads=num_threads,
                    env=METADATA[env][0],
                    label=f"x.zero_()  {dtype}",
                    description=repr(list(tensors['x'].size())),
                )
                results.append(timer.blocked_autorange(min_run_time=MIN_RUN_TIME))
                print(".", end="")
                sys.stdout.flush()

    counts = []
    print("\nStarting Callgrind measurements.")
    for setup in ["x = torch.empty(())", "x = torch.empty((1,))", "x = torch.empty((4, 4))[::2, ::2]"]:
        timer = Timer(
            "x.zero_()",
            setup=setup,
            env=METADATA[env][0],
        )
        counts.append(timer.collect_callgrind())
        print(".", end="")
        sys.stdout.flush()

    print("\nWriting results.")
    with open(METADATA[env][1], "wb") as f:
        pickle.dump([results, counts], f)


def compare_main():
    with open(METADATA["45636_ref"][1], "rb") as f:
        ref_results, ref_counts = pickle.load(f)

    with open(METADATA["45636_pr"][1], "rb") as f:
        pr_results, pr_counts = pickle.load(f)

    cmp = Compare(ref_results + pr_results)
    cmp.trim_significant_figures()
    cmp.colorize()
    cmp.print()

    print()
    for c_ref, c_pr in zip(ref_counts, pr_counts):
        for l_ref, l_pr in zip(repr(c_ref).splitlines(False), repr(c_pr).splitlines(False)):
            print(f"{l_ref[:70]:<75} |   {l_pr[:70]}")
        print("\n")


if __name__ == "__main__":
    if len(sys.argv) == 1:
        print("Pass `--measure` or `--compare`")
    elif sys.argv[1] == "--measure":
        measure_main()
    elif sys.argv[1] == "--compare":
        compare_main()
    else:
        print("Unknown arg(s):", sys.argv[1:])
```

From shell:
```
source activate 45636_ref
$ python benchmark_45636.py --measure
$ conda deactivate && source activate 45636_pr
$ python benchmark_45636.py --measure
$ python benchmark_45636.py --compare
```

<img width="1555" alt="Screen Shot 2020-10-02 at 1 27 00 AM" src="https://user-images.githubusercontent.com/13089297/94903314-77df1180-044e-11eb-905d-111c2f176c0a.png">
<img width="1545" alt="Screen Shot 2020-10-02 at 1 27 12 AM" src="https://user-images.githubusercontent.com/13089297/94903331-7c0b2f00-044e-11eb-9a7e-109afe5a5512.png">
<img width="1232" alt="Screen Shot 2020-10-02 at 1 51 57 AM" src="https://user-images.githubusercontent.com/13089297/94905537-e4a7db00-0451-11eb-8d2e-c69f61429426.png">

Results confirm @ngimel's observation that this would tank multi-threaded performance. I also took some instruction counts to gauge the effect that this would have on overheads. For the first two cases number of instructions drop by almost a factor of two which is also reflected in the wall times (no doubt because you're skipping a second round of dispatch); however that comes at the cost of a 1% increase in instruction overhead if we don't meet the special case of `self.device() == at::kCPU and self.is_contiguous()`.

Overall the performance when this approach works well makes me wonder if we can push this approach up into the `fill_` implementation, and then enjoy the benefits in a broader range of cases. WDYT?

DhavalKapil commented on aten/src/ATen/native/Fill.cpp:
```diff
@@ -106,7 +106,23 @@ Tensor& fill_diagonal_(Tensor& self, Scalar fill_value, bool wrap) {
+  int64_t nelements = at::prod_intlist(self.sizes());
+  int64_t size_bytes = nelements * self.dtype().itemsize();
+  if (size_bytes > 0) {
+    std::memset(ptr, 0, size_bytes);
```
We should avoid using `std::memset` to zero out buffers since it can be optimized away by the compiler :)
From https://en.cppreference.com/w/cpp/string/byte/memset:
> std::memset may be optimized away (under the as-if rules) if the object modified by this function is not accessed again for the rest of its lifetime (e.g. gcc bug 8537). For that reason, this function cannot be used to scrub memory (e.g. to fill an array that stored a password with zeroes). Solutions for that include **std::fill with volatile pointers**, **C11 memset_s**, **FreeBSD explicit_bzero** or **Microsoft SecureZeroMemory**.

ezyang replied:
Is gcc actually able to conclude that `ptr` becomes dead at the end of this function? I don't think so: it's owned by the tensor which escapes from this function.

edvgha commented:
@ngimel in TensorIterator there is special branch for for_each based on number of elements and number of threads.
https://github.com/pytorch/pytorch/blob/7731370e718ad16994d8da25fbdc9de64a22ec44/aten/src/ATen/native/TensorIterator.cpp#L694

Commit:
Optimize torch zeros (#45636)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45636

After creating empty tensor 'memset' used to zero out items of tensor

Test Plan:
pytorch benchmark tool results:

timer = benchmark_utils.Timer(stmt="torch.zeros((1024, 4096))")

Before: 1007 us
After:     841.26 us
1 measurement, 10000 runs , 1 thread

timer = benchmark_utils.Timer(stmt="torch.zeros((128))")

Before: 4 - 7.6 us
After:     2.4 - 2.8 us
1 measurement, 10000 runs , 1 thread

           torch.int8     |   1   |  4096  |  8192  |  16384  |  32768  |
1 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   500  |   600  |    700  |   2000  |
  (Reference)  x.zero_()  |  800  |  1000  |  1000  |   2000  |   2000  |
2 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   500  |   600  |    700  |   2000  |
  (Reference)  x.zero_()  |  800  |  1000  |  1000  |   2000  |   3000  |
4 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   500  |   600  |    700  |   2000  |
  (Reference)  x.zero_()  |  800  |  1000  |  1000  |   2000  |   3000  |

           torch.int32    |   1   |  4096  |  8192  |  16384  |  32768  |
1 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  400  |   700  |  2000  |   2900  |   5500  |
  (Reference)  x.zero_()  |  800  |  2000  |  3000  |   4400  |   7300  |
2 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   700  |  2000  |   3000  |   5600  |
  (Reference)  x.zero_()  |  900  |  2000  |  2000  |   3600  |   7200  |
4 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  400  |   700  |  2000  |   3000  |   5700  |
  (Reference)  x.zero_()  |  800  |  2000  |  3100  |   4300  |   9000  |

           torch.float16  |   1   |  4096  |  8192  |  16384  |  32768  |
1 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   500  |   700  |   2000  |   3000  |
  (Reference)  x.zero_()  |  800  |  1000  |  2000  |   2000  |   3300  |
2 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   600  |   700  |   2000  |   3000  |
  (Reference)  x.zero_()  |  800  |  1000  |  2000  |   2000  |   4300  |
4 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   600  |   700  |   2000  |   3300  |
  (Reference)  x.zero_()  |  900  |  1000  |  2000  |   2000  |   4400  |

           torch.float32  |   1   |  4096  |  8192  |  16384  |  32768  |
1 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   700  |  2000  |   3200  |   6100  |
  (Reference)  x.zero_()  |  800  |  2000  |  2000  |   3500  |   6100  |
2 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   700  |  2000  |   3100  |   5600  |
  (Reference)  x.zero_()  |  800  |  2000  |  2000  |   3300  |   7000  |
4 threads: --------------------------------------------------------------
  (PR #45636)  x.zero_()  |  500  |   700  |  2000  |   3000  |   5600  |
  (Reference)  x.zero_()  |  900  |  2000  |  2000  |   3600  |   7500  |

Differential Revision: D23925113

fbshipit-source-id: 97638a5f0a9913b37ad147b26d4928ce0701cb3e
